### DISCUSSION TITLE
Display matplotlib figure with v0.17.0a2

### DISCUSSION THREAD

**Discussion by jps1** [pyodide/pyodide#1392] [Q&A] [answered]
I've been playing around with pyodide for a project I'm working on, and had gotten some simple matplotlib scenarios working using version 0.16.1 by overwriting the iodide output element method as described [here](https://github.com/pyodide/pyodide/issues/479#issuecomment-549669229).  I decided to update to 0.17.0a2 and do my own build to include a couple third-party python libraries I'm using, and the build all went fine, but when I attempted to run the same matplotlib scenarios, I just got a blank webpage and no errors in the browser console.  After some digging around, I came across [this PR](https://github.com/pyodide/pyodide/pull/1057) which appears to have changed behavior such that the method I was using to add the div broke.

I then tried [this workaround](https://stackoverflow.com/a/59571016) to overwrite the FigureCanvasWasm create_root_element method to append the created div to document body, but ran into this:
![image](https://user-images.githubusercontent.com/8582709/113066444-62ce0b80-916f-11eb-9f4e-5f7b4843ee3c.png)
Some rudimentary debugging seems to indicate that the pixels byte array is just fine (size is correct, contents look okay), but maybe it's not being translated properly over to the javascript side of things (perhaps related to [this bug](https://github.com/pyodide/pyodide/issues/749)?).

I've pretty much reached the extent of my ability to easily trace what's going on here, so I'm hoping someone can help.  Am I approaching the whole matplotlib display thing wrong... should I be doing this some other way?  If that's the correct way, is the Uint8ClampedArray issue a bug and/or is there a recommended workaround?

Thanks!

**Comment by hoodmane** [comment, marked as answer]
Could you link your current code? (Maybe make a codepen?) The error you gave in the screenshot looks pretty helpful to me so I suspect I will be able to help you out.

Given that you have code that relies on the hack of defining an `iodide` object, what about just saying:
```python
from matplotlib.backends.wasm_backend import FigureCanvasWasm
def create_root_element(self):
    from js import iodide
    return iodide.output.element("div")
FigureCanvasWasm.create_root_element = create_root_element
```

**Reply by jps1** [reply]
I figured it was easier to just overwrite create_root_element() to do what I wanted, rather than overwriting create_root_element() to enable an old hack :-)  Also, I'd like to do things the recommended way as much as possible, since that tends to lead to less pain in the future!  If there's a better way to do this, I'm definitely open to suggestions for a more standard approach.

Here's a codepen of what I'm trying to do that replicates the same error:
https://codepen.io/jps1/pen/MWJJvvV

Also, not sure if it matters, but I'm on Firefox 86.0.1

**Reply by hoodmane** [reply]
> rather than overwriting create_root_element() to enable an old hack

Yeah it is pretty horrific. And it wouldn't even solve the problem.

I found the bug, will open a PR fixing this soon. Thanks for the report!

**Reply by hoodmane** [reply]
I got this working here (the only thing I changed in your code is the source to load Pyodide)
https://math.mit.edu/~hood/matplotlib-test/

It is currently blocked on PR #1322, but I'll open a PR on top of that soon.

**Reply by hoodmane** [reply]
Okay, this is PR #1414. If you build on that branch, matplotlib should work again.

**Reply by jps1** [reply]
Wonderful, thanks!  I'll give this a try in the next couple days and let you know how it goes.

**Reply by jps1** [reply]
I built using your matplotlib-backend branch, and it worked perfectly.  Thanks!
